lib/class-perfecty-push-lib-cron-check.php

```diff
@@ -0,0 +1,217 @@
+<?php
+
+use Perfecty_Push_Lib_Log as Log;
+
+/***
+ * Check cron functionality
+ */
+class Perfecty_Push_Lib_Cron_Check {
+
+	private const CRON_HOOK = 'perfecty_push_cron_check';
+
+	private const SCHEDULE_OFFSET = 1800; // 30 minutes
+
+	private const SUCCESS_RUN_LOG_MSG = 'Cron monitor runned';
+
+	private const NEXT_SCHED_LOG_MSG = 'Next cron check scheduled at: ';
+
+	private const NEVER_SCHED_LOG_MSG = 'Cron monitor was not scheduled';
+
+	private const NOT_RUNNED_LOG_MSG = 'Cron monitor was scheduled but not executed';
+
+	private const LIMIT_LOGS_FOR_CALC = 10;
+
+	/**
+	 * The cron running status.
+	 *
+	 * @since    1.2.0
+	 * @access   private
+	 * @var      bool    $cron_working    The cron running status.
+	 */
+	private static $cron_working = false;
+
+	/**
+	 * Initialize the class.
+	 *
+	 * @since    1.2.0
+	 */
+	public function __construct() {
+		$this->schedule_cron_job();
+	}
+
+	/**
+	 * Schedules the cron job once (this is the callback of the cronjob too).
+	 */
+	public function schedule_cron_job() {
+		if ( ! wp_next_scheduled( self::CRON_HOOK ) ) {
+			$now                 = time();
+			$next_scheduled_time = $now + self::SCHEDULE_OFFSET;
+			wp_schedule_single_event( $next_scheduled_time, self::CRON_HOOK );
+			Log::info(
+				self::SUCCESS_RUN_LOG_MSG . ' - ' . self::NEXT_SCHED_LOG_MSG . \
+				get_date_from_gmt( gmdate( 'Y-m-d H:i:s', intval( $next_scheduled_time ) ) ),
+				'Y-m-d H:i:s'
+			);
+		}
+	}
+
+	/**
+	 * Checks if next scheduled action is in the future and returns cron status
+	 *
+	 * @return bool
+	 */
+	public static function get_cron_status() {
+		$now = time();
+		/**
+		 * When wp-cron is disabled and a system cron is running we don't want to annoy users
+		 * with a false positive warning. In this scenario, a false positive (a warning saying
+		 * that no cron is set) could arise when the system's cron is set to call wp-cron.php
+		 * with a period (let's call it SYSCRON_PERIOD) longer than SCHEDULE_OFFSET.
+		 * We cannot directly read system cron settings, but we can guess the real periodicity
+		 * (given by the combination of the 2 periods) by reading our logs (when enabled).
+		 *
+		 * The REAL_PERIOD (we estimate) will be equal to the SYSCRON_PERIOD when this
+		 * is bigger (longer) than SCHEDULE_OFFSET.
+		 * If SYSCRON_PERIOD is smaller (shorter) than SCHEDULE_OFFSET it will be in range
+		 * [SCHEDULE_OFFSET; 2x SCHEDULE_OFFSET[
+		 *
+		 * The REAL_PERIOD can be calculated using this formula:
+		 *
+		 * IF ( SCHEDULE_OFFSET Mod ( SYSCRON_PERIOD ) > 0 ) {
+		 *      REAL_PERIOD = SCHEDULE_OFFSET + SYSCRON_PERIOD - ( SCHEDULE_OFFSET Mod ( SYSCRON_PERIOD ) )
+		 * } ELSE {
+		 *      REAL_PERIOD = SCHEDULE_OFFSET
+		 * }
+		 */
+		if ( ( ! self::is_wp_cron_enabled() ) && self::are_logs_enabled() ) {
+			$last_success_logs = self::get_success_logs();
```

Comment:
Logs are not part of the application logic, as they're  something that you throw and forget, so we should not depend on them. Also, the plugin could have a different Logger implementation, at the moment we support DB but later we could support others (syslog, error_log, ...).